AK/BuiltinWrappers.h

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2021, Nick Johnson <sylvyrfysh@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "Platform.h"
+
+template<typename IntType>
+inline constexpr int popcount(IntType value)
+{
+#if defined(__GNUC__) || defined(__clang__)
+    if constexpr (sizeof(IntType) <= 4)
+        return __builtin_popcount(value);
+    if constexpr (sizeof(IntType) == 8) {
+#    ifdef AK_OK_MACOS
+        return __builtin_popcountll(value);
+#    else
+        return __builtin_popcountl(value);
+#    endif
+    }
+#else
+    int ones = 0;
+    for (size_t i = 0; i < 8 * sizeof(IntType); ++i) {
+        if ((val >> i) & 1) {
+            ++ones;
+        }
+    }
+    return ones;
+#endif
+}
+
+template<typename IntType>
+inline constexpr int count_trailing_zeroes(IntType value)
+{
+#if defined(__GNUC__) || defined(__clang__)
+    if constexpr (sizeof(IntType) <= 4)
+        return __builtin_ctz(value);
+    if constexpr (sizeof(IntType) == 8) {
+#    ifdef AK_OK_MACOS
+        return __builtin_ctzll(value);
+#    else
+        return __builtin_ctzl(value);
+#    endif
+    }
+#else
+    for (size_t i = 0; i < 8 * sizeof(IntType); ++i) {
+        if ((val >> i) & 1) {
+            return i;
+        }
+    }
+    return 0;
+#endif
+}
+
+template<typename IntType>
+inline constexpr int count_trailing_zeroes_safe(IntType value)
+{
+    if (value == 0)
+        return 8 * sizeof(IntType);
+    return count_trailing_zeroes(value);
+}
+
+template<typename IntType>
+inline constexpr int count_leading_zeroes(IntType value)
+{
+#if defined(__GNUC__) || defined(__clang__)
+    if constexpr (sizeof(IntType) <= 4)
+        return __builtin_clz(value);
+    if constexpr (sizeof(IntType) == 8) {
+#    ifdef AK_OK_MACOS
+        return __builtin_clzll(value);
+#    else
+        return __builtin_clzl(value);
+#    endif
+    }
+#else
+    // Wrap around, catch going past zero by noticing that i is greater than the number of bits in the number
+    for (size_t i = (8 * sizeof(IntType)) - 1; i < 8 * sizeof(IntType); --i) {
+        if ((val >> i) & 1) {
+            return i;
+        }
+    }
+    return 0;
```

Comment:
The `__builtin_*` functions used everywhere have UD if `value == 0`, so I agree that making `count_leading_zeros()` return 0 when `value == 0` is okay.

However, do we really want to do it this way? If so, please add a comment that reminds the reader that this branch is never supposed to be taken anyway:
```
    // Although the value 0 has 8 * sizeof(IntType) leading zeros, the caller should have ensured that this function is never called with value 0.
```

(Same for `count_trailing_zeroes()`.)

Reply:
I copied the behavior from here https://github.com/SerenityOS/serenity/blob/master/AK/Platform.h#L110 , but I will comment this. I'll also check the `bit_scan_forward` implementation to make sure that this matched the correct behavior as well, thanks for pointing this out!

Reply:
Playing around with this, I found that all the builtins as well as the typical C++ standards require that the arguments be an unsigned type. As such, I'm adding those constraints and a test suite.

Reply:
Oh, it's a copy? Why keep the "old" copy in Platform.h? That sounds like it should be either removed, or replaced by an `#include <AK/BuiltinWrappers.h>`.

Reply:
I didn't want to modify all code that depended on that in this PR, but planned on removing everything else in a subsequent PR. Should I remove that and perform the necessary churn in a commit in here?

Reply:
Ah, if you're already working on that then I don't have a particular preference either way, as long as the code duplication is *eventually* removed. Maybe with the next PR, that's fine.